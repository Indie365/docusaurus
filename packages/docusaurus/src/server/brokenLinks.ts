/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import _ from 'lodash';
import logger from '@docusaurus/logger';
import {matchRoutes} from 'react-router-config';
import {parseURLPath, serializeURLPath, type URLPath} from '@docusaurus/utils';
import {getAllFinalRoutes} from './utils';
import type {RouteConfig, ReportingSeverity} from '@docusaurus/types';

type BrokenLink = {
  link: string;
  resolvedLink: string;
  anchor: boolean;
};

type BrokenLinksByLocation = {[location: string]: BrokenLink[]};

type CollectedLinks = {
  [location: string]: {links: string[]; anchors: string[]};
};

function getPageBrokenLinks({
  allCollectedLinks,
  pagePath,
  pageLinks,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  pagePath: string;
  pageLinks: string[];
  pageAnchors: string[];
  routes: RouteConfig[];
}): BrokenLink[] {
  // console.log('routes:', routes);
  function isPathBrokenLink(linkPath: URLPath) {
    const matchedRoutes = [linkPath.pathname, decodeURI(linkPath.pathname)]
      // @ts-expect-error: React router types RouteConfig with an actual React
      // component, but we load route components with string paths.
      // We don't actually access component here, so it's fine.
      .map((l) => matchRoutes(routes, l))
      .flat();
    return matchedRoutes.length === 0;
  }

  function isAnchorBrokenLink(linkPath: URLPath) {
    const {pathname, hash} = linkPath;

    // Link has no hash: it can't be a broken anchor link
    if (hash === undefined) {
      return false;
    }

    const targetPage =
      allCollectedLinks[pathname] || allCollectedLinks[decodeURI(pathname)];

    // link with anchor to a page that does not exist (or did not collect any
    // link/anchor) is considered as a broken anchor
    if (!targetPage) {
      return true;
    }

    // it's a broken anchor if the target page exists
    // but the anchor does not exist on that page
    return !targetPage.anchors.includes(hash);
  }

  const brokenLinks = pageLinks.flatMap((link) => {
    const linkPath = parseURLPath(link, pagePath);
    if (isPathBrokenLink(linkPath)) {
      return [
        {
          link,
          resolvedLink: serializeURLPath(linkPath),
          anchor: false,
        },
      ];
    }
    if (isAnchorBrokenLink(linkPath)) {
      return [
        {
          link,
          resolvedLink: serializeURLPath(linkPath),
          anchor: true,
        },
      ];
    }
    return [];
  });

  return brokenLinks;
}

/**
 * The route defs can be recursive, and have a parent match-all route. We don't
 * want to match broken links like /docs/brokenLink against /docs/*. For this
 * reason, we only consider the "final routes" that do not have subroutes.
 * We also need to remove the match-all 404 route
 */
function filterIntermediateRoutes(routesInput: RouteConfig[]): RouteConfig[] {
  const routesWithout404 = routesInput.filter((route) => route.path !== '*');
  return getAllFinalRoutes(routesWithout404);
}

function getAllBrokenLinks({
  allCollectedLinks,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  routes: RouteConfig[];
}): {brokenLinks: BrokenLinksByLocation; brokenAnchors: BrokenLinksByLocation} {
  const filteredRoutes = filterIntermediateRoutes(routes);

  const allBrokenLinks = _.mapValues(
    allCollectedLinks,
    (pageCollectedData, pagePath) =>
      getPageBrokenLinks({
        allCollectedLinks,
        pageLinks: pageCollectedData.links,
        pageAnchors: pageCollectedData.anchors,
        pagePath,
        routes: filteredRoutes,
      }),
  );

  function splitBrokenLinks(collect: {[x: string]: BrokenLink[]}) {
    const brokenLinks: {[x: string]: BrokenLink[]} = {};
    const brokenAnchors: {[x: string]: BrokenLink[]} = {};

    Object.entries(collect).forEach(([page, links]) => {
      const [linksFiltered, anchorsFiltered] = _.partition(
        links,
        (link) => link.anchor === false,
      );

      if (linksFiltered.length > 0) {
        brokenLinks[page] = linksFiltered;
      }
      if (anchorsFiltered.length > 0) {
        brokenAnchors[page] = anchorsFiltered;
      }
    });

    return {brokenLinks, brokenAnchors};
  }

  return splitBrokenLinks(allBrokenLinks);
}

function brokenLinkMessage(brokenLink: BrokenLink): string {
  const showResolvedLink = brokenLink.link !== brokenLink.resolvedLink;
  return `${brokenLink.link}${
    showResolvedLink ? ` (resolved as: ${brokenLink.resolvedLink})` : ''
  }`;
}

function createBrokenLinksMessage(
  pagePath: string,
  allBrokenLinks: BrokenLink[],
): string {
  const type = allBrokenLinks[0]?.anchor === true ? 'anchor' : 'link';

  const anchorMessage =
    allBrokenLinks.length > 0
      ? `- Broken ${type} on source page path = ${pagePath}:
   -> linking to ${allBrokenLinks
     .map(brokenLinkMessage)
     .join('\n   -> linking to ')}`
      : '';

  return `${anchorMessage}`;
}

function getAnchorBrokenLinksErrorMessage(allBrokenLinks: {
  [location: string]: BrokenLink[];
}): string | undefined {
  if (Object.keys(allBrokenLinks).length === 0) {
    return undefined;
  }

  return `Docusaurus found broken anchors!

Please check the pages of your site in the list below, and make sure you don't reference any anchor that does not exist.
Note: it's possible to ignore broken anchors with the 'onBrokenAnchors' Docusaurus configuration, and let the build pass.

Exhaustive list of all broken anchors found:
${Object.entries(allBrokenLinks)
  .map(([pagePath, brokenLinks]) =>
    createBrokenLinksMessage(pagePath, brokenLinks),
  )
  .join('\n')}
`;
}

function getPathBrokenLinksErrorMessage(allBrokenLinks: {
  [location: string]: BrokenLink[];
}): string | undefined {
  if (Object.keys(allBrokenLinks).length === 0) {
    return undefined;
  }

  /**
   * If there's a broken link appearing very often, it is probably a broken link
   * on the layout. Add an additional message in such case to help user figure
   * this out. See https://github.com/facebook/docusaurus/issues/3567#issuecomment-706973805
   */
  function getLayoutBrokenLinksHelpMessage() {
    const flatList = Object.entries(allBrokenLinks).flatMap(
      ([pagePage, brokenLinks]) =>
        brokenLinks.map((brokenLink) => ({pagePage, brokenLink})),
    );

    const countedBrokenLinks = _.countBy(
      flatList,
      (item) => item.brokenLink.link,
    );

    const FrequencyThreshold = 5; // Is this a good value?
    const frequentLinks = Object.entries(countedBrokenLinks)
      .filter(([, count]) => count >= FrequencyThreshold)
      .map(([link]) => link);

    if (frequentLinks.length === 0) {
      return '';
    }

    return logger.interpolate`

It looks like some of the broken links we found appear in many pages of your site.
Maybe those broken links appear on all pages through your site layout?
We recommend that you check your theme configuration for such links (particularly, theme navbar and footer).
Frequent broken links are linking to:${frequentLinks}`;
  }

  return `Docusaurus found broken links!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.${getLayoutBrokenLinksHelpMessage()}

Exhaustive list of all broken links found:
${Object.entries(allBrokenLinks)
  .map(([pagePath, brokenLinks]) =>
    createBrokenLinksMessage(pagePath, brokenLinks),
  )
  .join('\n')}
`;
}

export async function handleBrokenLinks({
  allCollectedLinks,
  onBrokenLinks,
  onBrokenAnchors,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  onBrokenLinks: ReportingSeverity;
  onBrokenAnchors: ReportingSeverity;
  routes: RouteConfig[];
}): Promise<void> {
  if (onBrokenLinks === 'ignore' && onBrokenAnchors === 'ignore') {
    return;
  }

  const {brokenLinks, brokenAnchors} = getAllBrokenLinks({
    allCollectedLinks,
    routes,
  });

  const pathErrorMessage = getPathBrokenLinksErrorMessage(brokenLinks);
  if (pathErrorMessage) {
    logger.report(onBrokenLinks)(pathErrorMessage);
  }

  const anchorErrorMessage = getAnchorBrokenLinksErrorMessage(brokenAnchors);
  if (anchorErrorMessage) {
    logger.report(onBrokenAnchors)(anchorErrorMessage);
  }
}
