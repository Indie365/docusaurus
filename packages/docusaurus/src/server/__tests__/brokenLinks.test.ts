/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {jest} from '@jest/globals';
import _ from 'lodash';
import {handleBrokenLinks} from '../brokenLinks';
import type {RouteConfig} from '@docusaurus/types';

describe('handleBrokenLinks', () => {
  const routes: RouteConfig[] = [
    {
      path: '/community',
      component: '',
    },
    {
      path: '/docs',
      component: '',
      routes: [
        {path: '/docs/goodDoc', component: ''},
        {path: '/docs/anotherGoodDoc', component: ''},
        {path: '/docs/good doc with space', component: ''},
        {path: '/docs/another good doc with space', component: ''},
        {path: '/docs/weird%20but%20good', component: ''},
      ],
    },
    {
      path: '*',
      component: '',
    },
  ];

  const link1 = '/link1';
  const link2 = '/docs/link2';
  const link3 = '/hey/link3';

  const allCollectedLinks = {
    '/docs/good doc with space': {
      links: [
        // Good - valid file with spaces in name
        './another%20good%20doc%20with%20space',
        // Good - valid file with percent-20 in its name
        './weird%20but%20good',
        // Bad - non-existent file with spaces in name
        './some%20other%20non-existent%20doc1',
        // Evil - trying to use ../../ but '/' won't get decoded
        // cSpell:ignore Fout
        './break%2F..%2F..%2Fout2',
      ],
      anchors: [],
    },
    '/docs/goodDoc': {
      links: [
        // Good links
        './anotherGoodDoc#someHash',
        '/docs/anotherGoodDoc?someQueryString=true#someHash',
        '../docs/anotherGoodDoc?someQueryString=true',
        '../docs/anotherGoodDoc#someHash',
        // Bad links
        '../anotherGoodDoc#reported-because-of-bad-relative-path1',
        './docThatDoesNotExist2',
        './badRelativeLink3',
        '../badRelativeLink4',
      ],
      anchors: [],
    },
    '/community': {
      links: [
        // Good links
        '/docs/goodDoc',
        '/docs/anotherGoodDoc#someHash',
        './docs/goodDoc#someHash',
        './docs/anotherGoodDoc',
        // Bad links
        '/someNonExistentDoc1',
        '/badLink2',
        './badLink3',
      ],
      anchors: [],
    },
    '/page1': {
      links: [link1],
      anchors: [],
    },
    '/page2': {
      links: [link2, link3],
      anchors: [],
    },
  };

  it('do not report anything for correct paths', async () => {
    const consoleMock = jest
      .spyOn(console, 'warn')
      .mockImplementation(() => {});
    const allCollectedCorrectLinks = {
      '/docs/good doc with space': {
        links: [
          './another%20good%20doc%20with%20space',
          './weird%20but%20good',
        ],
        anchors: [],
      },
      '/docs/goodDoc': {
        links: [
          '#someHash',
          '/community#anchorFromGoodDoc',
          './anotherGoodDoc#someHash',
          '/docs/anotherGoodDoc?someQueryString=true#someHash',
          '../docs/anotherGoodDoc?someQueryString=true',
          '../docs/anotherGoodDoc#someHash',
        ],
        anchors: ['someHash'], // anchors here are anchors of the page itself (/docs/goodDoc) not the links in it
      },
      '/community': {
        links: [
          '/docs/goodDoc',
          '/docs/anotherGoodDoc#someHash', // anchor here is an anchor of an other page, it should be checked against the anchors of the other page
          './docs/goodDoc#someHash',
          './docs/anotherGoodDoc',
        ],
        anchors: ['anchorFromGoodDoc'],
      },
    };
    await handleBrokenLinks({
      allCollectedLinks: allCollectedCorrectLinks,
      onBrokenLinks: 'warn',
      onBrokenAnchors: 'warn',
      routes,
    });
    expect(consoleMock).toHaveBeenCalledTimes(0);
  });

  it('reports all broken links', async () => {
    await expect(() =>
      handleBrokenLinks({
        allCollectedLinks,
        onBrokenLinks: 'throw',
        onBrokenAnchors: 'throw',
        routes,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('no-op for ignore', async () => {
    // TODO this mock is not future-proof, we may remove mapValues
    // In any case, _.mapValues will always be called, unless handleBrokenLinks
    // has already bailed
    const lodashMock = jest.spyOn(_, 'mapValues');
    await handleBrokenLinks({
      allCollectedLinks,
      onBrokenLinks: 'ignore',
      onBrokenAnchors: 'ignore',
      routes,
    });
    expect(lodashMock).toHaveBeenCalledTimes(0);
    lodashMock.mockRestore();
  });

  it('reports frequent broken links', async () => {
    Object.values(allCollectedLinks).forEach(({links}) => {
      links.push(
        '/frequent',
        // This is in the gray area of what should be reported. Relative paths
        // may be resolved to different slugs on different locations. But if
        // this comes from a layout link, it should be reported anyways
        './maybe-not',
      );
    });

    await expect(() =>
      handleBrokenLinks({
        allCollectedLinks,
        onBrokenLinks: 'throw',
        onBrokenAnchors: 'throw',
        routes,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
