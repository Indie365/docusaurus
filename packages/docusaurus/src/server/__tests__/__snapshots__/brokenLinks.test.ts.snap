// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBrokenLinks reports all broken links 1`] = `
"Docusaurus found broken links / anchors!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' or 'onBrokenAnchors' Docusaurus configuration, and let the build pass.

Exhaustive list of all broken links found:
- Broken link on source page path = /docs/good doc with space:
   -> linking to ./some%20other%20non-existent%20doc1 (resolved as: /docs/some%20other%20non-existent%20doc1)
   -> linking to ./break%2F..%2F..%2Fout2 (resolved as: /docs/break%2F..%2F..%2Fout2)

- Broken link on source page path = /docs/goodDoc:
   -> linking to ../anotherGoodDoc#reported-because-of-bad-relative-path1 (resolved as: /anotherGoodDoc)
   -> linking to ./docThatDoesNotExist2 (resolved as: /docs/docThatDoesNotExist2)
   -> linking to ./badRelativeLink3 (resolved as: /docs/badRelativeLink3)
   -> linking to ../badRelativeLink4 (resolved as: /badRelativeLink4)

- Broken link on source page path = /community:
   -> linking to /someNonExistentDoc1
   -> linking to /badLink2
   -> linking to ./badLink3 (resolved as: /badLink3)

- Broken link on source page path = /page1:
   -> linking to /link1
   -> linking to /files/hey.html
   -> linking to /javadoc
   -> linking to /files/hey
   -> linking to /javadoc/index.html
   -> linking to /javadoc/index.html#foo (resolved as: /javadoc/index.html)
   -> linking to /emptyFolder

- Broken link on source page path = /page2:
   -> linking to /docs/link2
   -> linking to /emptyFolder/
   -> linking to /javadoc/
   -> linking to /hey/link3
   -> linking to /javadoc/index.html
   -> linking to /files/file.zip

"
`;

exports[`handleBrokenLinks reports frequent broken links 1`] = `
"Docusaurus found broken links / anchors!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' or 'onBrokenAnchors' Docusaurus configuration, and let the build pass.

It looks like some of the broken links we found appear in many pages of your site.
Maybe those broken links appear on all pages through your site layout?
We recommend that you check your theme configuration for such links (particularly, theme navbar and footer).
Frequent broken links are linking to:
- /frequent
- ./maybe-not

Exhaustive list of all broken links found:
- Broken link on source page path = /docs/good doc with space:
   -> linking to ./some%20other%20non-existent%20doc1 (resolved as: /docs/some%20other%20non-existent%20doc1)
   -> linking to ./break%2F..%2F..%2Fout2 (resolved as: /docs/break%2F..%2F..%2Fout2)
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /docs/maybe-not)

- Broken link on source page path = /docs/goodDoc:
   -> linking to ../anotherGoodDoc#reported-because-of-bad-relative-path1 (resolved as: /anotherGoodDoc)
   -> linking to ./docThatDoesNotExist2 (resolved as: /docs/docThatDoesNotExist2)
   -> linking to ./badRelativeLink3 (resolved as: /docs/badRelativeLink3)
   -> linking to ../badRelativeLink4 (resolved as: /badRelativeLink4)
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /docs/maybe-not)

- Broken link on source page path = /community:
   -> linking to /someNonExistentDoc1
   -> linking to /badLink2
   -> linking to ./badLink3 (resolved as: /badLink3)
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /maybe-not)

- Broken link on source page path = /page1:
   -> linking to /link1
   -> linking to /files/hey.html
   -> linking to /javadoc
   -> linking to /files/hey
   -> linking to /javadoc/index.html
   -> linking to /javadoc/index.html#foo (resolved as: /javadoc/index.html)
   -> linking to /emptyFolder
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /maybe-not)

- Broken link on source page path = /page2:
   -> linking to /docs/link2
   -> linking to /emptyFolder/
   -> linking to /javadoc/
   -> linking to /hey/link3
   -> linking to /javadoc/index.html
   -> linking to /files/file.zip
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /maybe-not)

"
`;
